xinference/model/llm/ggml/ctransformers_util.py

```diff
@@ -0,0 +1,165 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+import re
+import time
+import uuid
+from typing import Iterator, Optional, Sequence, Tuple
+
+from ....types import CompletionChoice, CompletionChunk, CompletionUsage
+
+logger = logging.getLogger(__name__)
+
+
+def generate_stream(
+    model,
+    model_ref,
+    prompt: str,
+    *,
+    max_new_tokens: Optional[int] = None,
+    top_k: Optional[int] = None,
+    top_p: Optional[float] = None,
+    temperature: Optional[float] = None,
+    repetition_penalty: Optional[float] = None,
+    last_n_tokens: Optional[int] = None,
+    seed: Optional[int] = None,
+    batch_size: Optional[int] = None,
+    stream: Optional[bool] = False,
+    threads: Optional[int] = None,
+    stop: Optional[Sequence[str]] = None,
+    reset: Optional[bool] = None,
+    **kwargs,
+) -> Iterator[Tuple[CompletionChunk, CompletionUsage]]:
+    stop = stop or []
+    if isinstance(stop, str):
+        stop = [stop]
+
+    tokens = model_ref.tokenize(prompt)
+
+    stop_regex = re.compile("|".join(map(re.escape, stop)))
+    count = 0
+    text = ""
+    total_text = ""
+    incomplete = b""
+
+    # parameters needed for Xinference.
+    finish_reason = None
+
+    for token in model_ref.generate(
+        tokens,
+        top_k=top_k,
+        top_p=top_p,
+        temperature=temperature,
+        repetition_penalty=repetition_penalty,
+        last_n_tokens=last_n_tokens,
+        seed=seed,
+        batch_size=batch_size,
+        threads=threads,
+        reset=reset,
+    ):
+        # Handle incomplete UTF-8 multi-byte characters.
+        try:
+            from ctransformers.utils import utf8_split_incomplete
+        except ImportError:
+            error_message = (
+                "Failed to import module 'ctransformers - utf8_split_incomplete'"
+            )
+
+            installation_guide = [
+                "Please make sure 'ctransformers' is installed. You can install it by checking out the repository: "
+                "https://github.com/marella/ctransformers",
+            ]
+
+            raise ImportError(f"{error_message}\n\n{''.join(installation_guide)}")
+
+        incomplete += model_ref.detokenize([token], decode=False)
+        complete, incomplete = utf8_split_incomplete(incomplete)
+        output = complete.decode(errors="ignore")
+        text += output
+        total_text += output
+
+        logger.debug("Output, completion: %s", text)
```

Comment:
```suggestion
```

Don't need to log for every generated token.

Reply:
removed in the latest version